src/gatsby-node.js

```diff
@@ -155,17 +156,25 @@ const serializeSources = ({ mapping, additionalSitemaps = [] }) => {
     return sitemaps
 }
 
-const runQuery = (handler, { query, exclude }) => handler(query).then((r) => {
+const runQuery = (handler, { query, mapping, exclude }) => handler(query).then((r) => {
     if (r.errors) {
         throw new Error(r.errors.join(`, `))
     }
 
     for (let source in r.data) {
+        // Check for custom serializer
+        if (mapping && mapping[source] && typeof mapping[source].customSerializer === `function`) {
+            if (r.data[source] && r.data[source].edges && r.data[source].edges.length) {
+                r.data[source].edges = mapping[source].customSerializer(r.data[source].edges)
+            }
+        }
+
         // Removing excluded paths
         if (r.data[source] && r.data[source].edges && r.data[source].edges.length) {
             r.data[source].edges = r.data[source].edges.filter(({ node }) => !exclude.some((excludedRoute) => {
                 const sourceType = node.__typename ? `all${node.__typename}` : source
-                const slug = (sourceType === `allMarkdownRemark` || sourceType === `allMdx`) ? node.fields.slug.replace(/^\/|\/$/, ``) : node.slug.replace(/^\/|\/$/, ``)
+                const slug = (sourceType === `allMarkdownRemark` || sourceType === `allMdx`) || (node.fields && node.fields.slug) ? node.fields.slug.replace(/^\/|\/$/, ``) : node.slug.replace(/^\/|\/$/, ``)
```

Comment:
I rebased this with current master. Could you verify that this still works for your use case?